### PR TITLE
Add Azure Firewall Premium Templates

### PR DESCRIPTION
**Pull request by gersner**
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog
Addition of new Azure Firewall Premium features

### CODE REVIEW

File: 101-azurefirewall-premium/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Create Firewall Premium (IDPS,TLS Inspection,Web Categories)
```
Comment (bmoore-msft) [inline 101-azurefirewall-premium/README.md:1]: see pipeline results for changes needed here

File: 101-azurefirewall-premium/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    "variables": {
+        "firewallAddressSpace": "10.0.100.0/24",
+        "firewallPrivateIPAddress": "10.0.100.4",
+        "serverAddressSpace": "10.0.10.0/24",
+        "keyVaultName": "[concat('fw-demo-', uniqueString(resourceGroup().id))]",
+        "keyVaultCASecretName": "CACert"
+    },
+    "resources": [
+        {
+            "apiVersion": "2020-05-01",
```
Comment (bmoore-msft) [inline 101-azurefirewall-premium/azuredeploy.json:22]: See: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#sort-order-of-properties for the proper sort order of properties

File: 101-azurefirewall-premium/prereqs/.settings.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (bmoore-msft) [inline 101-azurefirewall-premium/prereqs/.settings.json:1]: I don't see anything that suggests the prereqs need to be in the same resource group - so you can remove this file.   you'll need to sure your resource references include the resourceGroup as well.

File: 101-azurefirewall-premium/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+    "parameters": {
+        "location": {
+            "type": "string",
+            "defaultValue": "[resourceGroup().location]",
+            "metadata": {
+                "description": "Deployment location"
+            }
+        }
+    },
+    "variables": {
```
Comment (bmoore-msft) [inline 101-azurefirewall-premium/azuredeploy.json:13]: it looks like most of these variables need to be parameters, as they are affected by settings in the prereqs.  you will also need to add the resourceGroup name for the keyvault and the names of vnets/subnets

File: 101-azurefirewall-premium/prereqs/prereq.azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+        "vmSize": {
+            "type": "string",
+            "defaultValue": "Standard_B2s",
+            "metadata": {
+                "description": "Size of virtual machine sizes"
+            }
+        },
+        "remoteAccessUsername": {
+            "type": "string",
+            "defaultValue": "AzureUser",
```
Comment (bmoore-msft) [inline 101-azurefirewall-premium/prereqs/prereq.azuredeploy.json:14]: remove defaultValue

File: 101-azurefirewall-premium/prereqs/prereq.azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        {
+            "apiVersion": "2019-09-01",
+            "type": "Microsoft.KeyVault/vaults",
+            "name": "[variables('keyVaultName')]",
+            "location": "[parameters('location')]",
+            "dependsOn": [
+                "[resourceId('Microsoft.ManagedIdentity/userAssignedIdentities', 'DemoIdentity')]"
+            ],
+            "properties": {
+                "enabledForDeployment": true,
```
Comment (bmoore-msft) [inline 101-azurefirewall-premium/prereqs/prereq.azuredeploy.json:132]: why is this needed?

File: 101-azurefirewall-premium/prereqs/prereq.azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+                "enabledForDiskEncryption": false,
+                "enabledForTemplateDeployment": true,
```
Comment (bmoore-msft) [inline 101-azurefirewall-premium/prereqs/prereq.azuredeploy.json:134]: why is this needed?

File: 101-azurefirewall-premium/prereqs/prereq.azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+                            "publicIPAddress": {
+                                "id": "[resourceId('Microsoft.Network/publicIpAddresses', 'BastionPublicIP')]"
+                            }
+                        }
+                    }
+                ]
+            }
+        }
+    ]
+}
```
Comment (bmoore-msft) [inline 101-azurefirewall-premium/prereqs/prereq.azuredeploy.json:309]: you'll need to add outputs here for the parameters that are added to azuredeploy.json

**Review by bmoore-msft — CHANGES_REQUESTED** [review]

File: 101-azurefirewall-premium/README.md (hunk 1 above)
Comment (gersner) [reply 101-azurefirewall-premium/README.md:1]: @bmoore-msft are you referring to the Fairfax badge? Is that needed if we do not support fairfax?

File: 101-azurefirewall-premium/prereqs/prereq.azuredeploy.json (hunk 6 above)
Comment (gersner) [reply 101-azurefirewall-premium/prereqs/prereq.azuredeploy.json:132]: Earlier iteration. Removed.

File: 101-azurefirewall-premium/prereqs/prereq.azuredeploy.json (hunk 8 above)
Comment (gersner) [reply 101-azurefirewall-premium/prereqs/prereq.azuredeploy.json:309]: Moving to single azuredeploy

File: 101-azurefirewall-premium/prereqs/prereq.azuredeploy.json (hunk 5 above)
Comment (gersner) [reply 101-azurefirewall-premium/prereqs/prereq.azuredeploy.json:14]: Ack

File: 101-azurefirewall-premium/azuredeploy.json (hunk 4 above)
Comment (gersner) [reply 101-azurefirewall-premium/azuredeploy.json:13]: Moving to single azuredeploy.

File: 101-azurefirewall-premium/azuredeploy.json (hunk 2 above)
Comment (gersner) [reply 101-azurefirewall-premium/azuredeploy.json:22]: Thanks

**Comment by gersner** [comment]
Hi @bmoore-msft, can you assist with reviewing this PR?
